File: integrations/python/zep_crewai/src/zep_crewai/tools.py
"""
Zep CrewAI Tools.

This module provides CrewAI tools for interacting with Zep memory storage,
including graph and user memory operations.
"""

import logging
from typing import Any

from crewai.tools import BaseTool
from pydantic import BaseModel, Field
from zep_cloud.client import Zep

logger = logging.getLogger(__name__)


class SearchMemoryInput(BaseModel):
    """Input schema for memory search tool."""

    query: str = Field(..., description="The search query to find relevant memories")
    limit: int = Field(default=10, description="Maximum number of results to return")
    scope: str = Field(
        default="edges",
        description="Scope of search: 'edges' (facts), 'nodes' (entities), 'episodes', or 'all'",
    )


class AddGraphDataInput(BaseModel):
    """Input schema for adding data to graph."""

    data: str = Field(..., description="The data/information to store in the graph")
    data_type: str = Field(default="text", description="Type of data: 'text', 'json', or 'message'")


class ZepSearchTool(BaseTool):
    """
    Tool for searching Zep memory storage.

    Can search either graph memory or user memory depending on initialization.
    """

    name: str = "Zep Memory Search"
    description: str = "Search Zep memory storage for relevant information"
    args_schema: type[BaseModel] = SearchMemoryInput

    def __init__(
        self, client: Zep, graph_id: str | None = None, user_id: str | None = None, **kwargs: Any
    ):
        """
        Initialize search tool bound to either a graph or user.

        Args:
            client: Zep client instance
            graph_id: Graph ID for generic knowledge graph search
            user_id: User ID for user-specific graph search
            **kwargs: Additional configuration
        """
        if not graph_id and not user_id:
            raise ValueError("Either graph_id or user_id must be provided")

        if graph_id and user_id:
            raise ValueError("Only one of graph_id or user_id should be provided")

        # Update description based on target
        if graph_id:
            kwargs["description"] = f"Search Zep graph '{graph_id}' for relevant information"
        else:
            kwargs["description"] = f"Search user '{user_id}' memories for relevant information"

        super().__init__(**kwargs)

        # Store as private attributes to avoid Pydantic validation
        self._client = client
        self._graph_id = graph_id
        self._user_id = user_id

    @property
    def client(self) -> Zep:
        """Get the Zep client."""
        return self._client

    @property
    def graph_id(self) -> str | None:
        """Get the graph ID."""
        return self._graph_id

    @property
    def user_id(self) -> str | None:
        """Get the user ID."""
        return self._user_id

    def _run(self, query: str, limit: int = 10, scope: str = "edges") -> str:
        """
        Execute the search operation.

        Args:
            query: Search query
            limit: Maximum results
            scope: Search scope

        Returns:
            Formatted search results
        """
        try:
            results = []

            if scope == "all":
                # Search all scopes
                scopes = ["edges", "nodes", "episodes"]
            else:
                scopes = [scope]

            for search_scope in scopes:
                if self._graph_id:
                    # Search graph memory
                    search_results = self._client.graph.search(
                        graph_id=self._graph_id, query=query, limit=limit, scope=search_scope
                    )
                else:
                    # Search user memory
                    search_results = self._client.graph.search(
                        user_id=self._user_id, query=query, limit=limit, scope=search_scope
                    )

                # Process results based on scope
                if search_scope == "edges" and search_results.edges:
                    for edge in search_results.edges:
                        results.append(
                            {
                                "type": "fact",
                                "content": edge.fact,
                                "name": edge.name,
                                "created_at": str(edge.created_at) if edge.created_at else None,
                            }
                        )

                elif search_scope == "nodes" and search_results.nodes:
                    for node in search_results.nodes:
                        results.append(
                            {
                                "type": "entity",
                                "content": f"{node.name}: {node.summary}",
                                "name": node.name,
                                "created_at": str(node.created_at) if node.created_at else None,
                            }
                        )

                elif search_scope == "episodes" and search_results.episodes:
                    for episode in search_results.episodes:
                        results.append(
                            {
                                "type": "episode",
                                "content": episode.content,
                                "source": episode.source,
                                "role": episode.role,
                                "created_at": str(episode.created_at)
                                if episode.created_at
                                else None,
                            }
                        )

            if not results:
                return f"No results found for query: '{query}'"

            # Format results for agent consumption
            formatted = f"Found {len(results)} relevant memories:\n\n"
            for i, result in enumerate(results, 1):
                result_type = result.get("type", "unknown")
                formatted += f"{i}. [{result_type.upper() if result_type else 'UNKNOWN'}] {result['content']}\n"
                if result.get("created_at"):
                    formatted += f"   (Created: {result['created_at']})\n"
                formatted += "\n"

            logger.info(f"Found {len(results)} memories for query: {query}")
            return formatted

        except Exception as e:
            error_msg = f"Error searching Zep memory: {str(e)}"
            logger.error(error_msg)
            return error_msg


class ZepAddDataTool(BaseTool):
    """
    Tool for adding data to Zep memory storage.

    Can add data to either graph memory or user memory depending on initialization.
    """

    name: str = "Zep Add Data"
    description: str = "Add data to Zep memory storage"
    args_schema: type[BaseModel] = AddGraphDataInput

    def __init__(
        self, client: Zep, graph_id: str | None = None, user_id: str | None = None, **kwargs: Any
    ):
        """
        Initialize add data tool bound to either a graph or user.

        Args:
            client: Zep client instance
            graph_id: Graph ID for generic knowledge graph
            user_id: User ID for user-specific graph
            **kwargs: Additional configuration
        """
        if not graph_id and not user_id:
            raise ValueError("Either graph_id or user_id must be provided")

        if graph_id and user_id:
            raise ValueError("Only one of graph_id or user_id should be provided")

        # Update description based on target
        if graph_id:
            kwargs["description"] = f"Add data to Zep graph '{graph_id}'"
        else:
            kwargs["description"] = f"Add data to user '{user_id}' memory"

        super().__init__(**kwargs)

        # Store as private attributes to avoid Pydantic validation
        self._client = client
        self._graph_id = graph_id
        self._user_id = user_id

    @property
    def client(self) -> Zep:
        """Get the Zep client."""
        return self._client

    @property
    def graph_id(self) -> str | None:
        """Get the graph ID."""
        return self._graph_id

    @property
    def user_id(self) -> str | None:
        """Get the user ID."""
        return self._user_id

    def _run(self, data: str, data_type: str = "text") -> str:
        """
        Execute the add data operation.

        Args:
            data: Data to store
            data_type: Type of data

        Returns:
            Success or error message
        """
        try:
            # Validate data type
            if data_type not in ["text", "json", "message"]:
                data_type = "text"

            if self._graph_id:
                # Add to graph memory
                self._client.graph.add(graph_id=self._graph_id, type=data_type, data=data)

                success_msg = f"Successfully added {data_type} data to graph '{self._graph_id}'"
                logger.debug(f"Added data to graph {self._graph_id}: {data[:100]}...")

            else:
                # Add to user graph memory
                self._client.graph.add(user_id=self._user_id, type=data_type, data=data)

                success_msg = (
                    f"Successfully added {data_type} data to user '{self._user_id}' memory"
                )
                logger.debug(f"Added data to user {self._user_id}: {data[:100]}...")

            return success_msg

        except Exception as e:
            error_msg = f"Error adding data to Zep: {str(e)}"
            logger.error(error_msg)
            return error_msg


def create_search_tool(
    client: Zep, graph_id: str | None = None, user_id: str | None = None
) -> ZepSearchTool:
    """
    Create a search tool bound to a Zep client.

    Args:
        client: Zep client instance
        graph_id: Optional graph ID for generic knowledge graph
        user_id: Optional user ID for user-specific graph

    Returns:
        ZepSearchTool instance

    Raises:
        ValueError: If neither or both IDs are provided
    """
    return ZepSearchTool(client=client, graph_id=graph_id, user_id=user_id)


def create_add_data_tool(
    client: Zep, graph_id: str | None = None, user_id: str | None = None
) -> ZepAddDataTool:
    """
    Create an add data tool bound to a Zep client.

    Args:
        client: Zep client instance
        graph_id: Optional graph ID for generic knowledge graph
        user_id: Optional user ID for user-specific graph

    Returns:
        ZepAddDataTool instance

    Raises:
        ValueError: If neither or both IDs are provided
    """
    return ZepAddDataTool(client=client, graph_id=graph_id, user_id=user_id)

File: integrations/vercel-ai/typescript/src/tools.ts
import { tool } from "ai";
import type { ToolSet } from "ai";
import { z } from "zod";
import type { ZepClient, Zep } from "@getzep/zep-cloud";
import type { ZepBinding, ZepLogger } from "./types.js";
import {
  GRAPH_MAX_CHARS,
  MESSAGE_MAX_CHARS,
  errorMessage,
  resolveGraphTarget,
  resolveLogger,
  toRoleType,
  truncateForZep,
} from "./zep-utils.js";

/** Options shared by every tool factory. */
interface BaseToolOptions {
  /** A shared, initialized Zep client. The caller owns its lifecycle. */
  client: ZepClient;
  /** Logger for Zep failures. Defaults to `console`. */
  logger?: ZepLogger;
}

/**
 * Every search parameter that can be pinned or exposed to the model, by name.
 * `scope` selects the Zep v4 search method (`graph.searchEdges`,
 * `graph.searchNodes`, and the other scope methods). The remaining keys match
 * the camelCase fields of the SDK's `SearchRequest`, except `limit`, which is
 * the page size of the search request.
 */
export type ZepSearchParamName = "scope" | "reranker" | "limit" | "mmrLambda" | "centerNodeUuid";

/** The search scopes this tool supports. Each scope selects one v4 method. */
export type ZepSearchScope = (typeof SCOPE_VALUES)[number];

/** Options for {@link createZepSearchTool}. */
export interface ZepSearchToolOptions extends BaseToolOptions {
  /** The graph to search, addressed by its UUID. */
  binding: ZepBinding;
  /** Override the tool description shown to the model. */
  description?: string;
  /**
   * Pin a search parameter to a fixed value: hidden from the model's tool
   * schema and always sent with the given value, regardless of what the model
   * would otherwise choose.
   */
  pinnedParams?: Partial<Record<ZepSearchParamName, unknown>>;
  /**
   * Hide a search parameter from the model's tool schema WITHOUT pinning it —
   * the parameter is simply omitted from the SDK call, so Zep's own
   * server-side default applies.
   */
  hiddenParams?: Set<ZepSearchParamName> | ZepSearchParamName[];
  /**
   * Optional Zep search filters (entity/edge types, properties, dates).
   * Constructor-only — never exposed to the model.
   */
  filters?: Zep.SearchFilters;
  /**
   * Node UUIDs seeding a breadth-first search. Constructor-only — never
   * exposed to the model.
   */
  bfsOriginNodeUuids?: string[];
  /**
   * Deprecated back-compat alias for `pinnedParams.scope`. Defaults to
   * `"edges"` remain the model's default when this and `pinnedParams.scope`
   * are both unset — the parameter stays exposed.
   */
  scope?: ZepSearchScope;
  /** Deprecated back-compat alias for `pinnedParams.limit`. */
  limit?: number;
  /** Deprecated back-compat alias for `pinnedParams.reranker`. */
  reranker?: Zep.V4SearchRequestReranker;
}

/** Options for {@link createZepRememberTool}. */
export interface ZepRememberToolOptions extends BaseToolOptions {
  /**
   * The graph this tool writes to, plus the thread used to record conversation.
   *
   * - When `threadUuid` is present, conversational content (a `role` is
   *   supplied) is persisted via `thread.addMessages` (records history *and*
   *   ingests into the user graph). Non-conversational data always uses
   *   `graph.episode.add`.
   * - When no thread is bound, everything is ingested via `graph.episode.add`.
   */
  binding: ZepBinding & { threadUuid?: string };
  /** Override the tool description shown to the model. */
  description?: string;
  /** Default `name` recorded on conversational messages (e.g. the user's real name). */
  defaultMessageName?: string;
}

/** Options for {@link createZepContextTool}. */
export interface ZepContextToolOptions extends BaseToolOptions {
  /**
   * The thread whose Context Block to fetch. The block is assembled from the
   * **entire user graph**; the thread only scopes relevance.
   */
  threadUuid: string;
  /** Override the tool description shown to the model. */
  description?: string;
  /** Optional UUID of a Zep context template for custom Context Block formatting. */
  templateUuid?: string;
}

/** The search scopes this tool supports, exposed to the model. */
const SCOPE_VALUES = [
  "edges",
  "nodes",
  "episodes",
  "observations",
  "thread_summaries",
  "auto",
] as const;

/** Zep's supported rerankers (`V4SearchRequestReranker`). */
const RERANKER_VALUES = ["rrf", "mmr", "node_distance", "episode_mentions", "cross_encoder"] as const;

const DEFAULT_SEARCH_SCOPE: ZepSearchScope = "edges";

/** Zep's server-side ceiling for the search result `limit`. */
const MAX_SEARCH_LIMIT = 50;

/**
 * Rerankers Zep rejects outright when `scope` is `"auto"` (auto search always
 * uses RRF internally and ignores `reranker` entirely).
 */
const AUTO_INCOMPATIBLE_RERANKERS: ReadonlySet<string> = new Set([
  "node_distance",
  "episode_mentions",
]);

/**
 * Clamp a search limit into Zep's accepted range `[1, 50]` so the call never
 * 400s. Warns (via `warn`) only when the limit exceeds the ceiling.
 */
function clampSearchLimit(limit: number, warn: (message: string) => void): number {
  if (limit > MAX_SEARCH_LIMIT) {
    warn(
      `[zep-search] limit ${limit} exceeds Zep ceiling ${MAX_SEARCH_LIMIT}; ` +
        `clamping to ${MAX_SEARCH_LIMIT}.`,
    );
    return MAX_SEARCH_LIMIT;
  }
  return limit < 1 ? 1 : limit;
}

const queryField = z
  .string()
  .min(1)
  .max(400)
  .describe(
    "What to look up in long-term memory (max 400 characters). Phrase it as " +
      "the information you need, e.g. 'where the user lives'.",
  );

/**
 * Zod field builders for each pin-or-exposable search parameter,
 * matching the pydantic-ai sibling's `_SEARCH_PARAM_SPECS` shape (description,
 * default) so the model-facing contract stays consistent across languages.
 */
const SEARCH_FIELD_BUILDERS: Record<ZepSearchParamName, () => z.ZodTypeAny> = {
  scope: () =>
    z
      .enum(SCOPE_VALUES)
      .optional()
      .describe(
        "What to search for: 'edges' for facts and relationships, " +
          "'nodes' for entities and their summaries, " +
          "'episodes' for raw text data (unstructured text, messages, or JSON), " +
          "'observations' for derived memories, " +
          "'thread_summaries' for incremental thread summaries, " +
          "'auto' to let Zep decide the best mix of results. Defaults to 'edges'.",
      ),
  reranker: () =>
    z
      .enum(RERANKER_VALUES)
      .optional()
      .describe(
        "Result ordering algorithm: 'rrf' (balanced), 'mmr' (diverse), " +
          "'cross_encoder' (highest accuracy), 'episode_mentions' " +
          "(frequently referenced), 'node_distance' (near a specific entity). Defaults to 'rrf'.",
      ),
  limit: () =>
    z.number().int().optional().describe("Maximum number of results to return. Defaults to 10."),
  mmrLambda: () =>
    z
      .number()
      .optional()
      .describe(
        "Balance between diversity (0.0) and relevance (1.0). Only used when reranker is 'mmr'.",
      ),
  centerNodeUuid: () =>
    z
      .string()
      .optional()
      .describe(
        "UUID of the center node for distance-based reranking. Required when reranker is 'node_distance'.",
      ),
};

/**
 * Build the model-facing zod input schema for the search tool, excluding
 * pinned/hidden parameters. `query` is always present and required.
 */
function buildSearchInputSchema(pinned: Set<ZepSearchParamName>, hidden: Set<ZepSearchParamName>) {
  const shape: Record<string, z.ZodTypeAny> = { query: queryField };
  for (const name of Object.keys(SEARCH_FIELD_BUILDERS) as ZepSearchParamName[]) {
    if (pinned.has(name) || hidden.has(name)) continue;
    shape[name] = SEARCH_FIELD_BUILDERS[name]();
  }
  return z.object(shape);
}

const rememberInputSchema = z.object({
  content: z
    .string()
    .min(1)
    .describe("The fact, message, or piece of information to remember."),
  role: z
    .string()
    .optional()
    .describe(
      "Who the content is from when it is a conversational message: 'user', " +
        "'assistant', 'system', 'tool', or 'function'. Omit for general facts.",
    ),
});

const contextInputSchema = z.object({});

/** Format a node-like result ("name: summary" or just "name"). */
function nameAndSummary(n: { name?: string; summary?: string }): string | undefined {
  if (!n.name) return undefined;
  return n.summary ? `${n.name}: ${n.summary}` : n.name;
}

const isNonEmpty = (s: string | undefined): s is string => Boolean(s);

/**
 * Run the Zep v4 search method for the active scope and extract
 * human-readable strings from the first page of results.
 *
 * Zep v4 has one method for each scope instead of a single `graph.search` call
 * with a `scope` field. Scope `auto` has no search method; it maps to
 * `graph.getContext`, which assembles one context block across every context
 * type. The switch is exhaustive over {@link ZepSearchScope}; the `never`
 * default makes a new scope a compile error rather than a silently-empty
 * result.
 */
async function searchByScope(
  client: ZepClient,
  graphUuid: string,
  scope: ZepSearchScope,
  request: { limit?: number; body: Zep.SearchRequest },
): Promise<string[]> {
  switch (scope) {
    case "auto": {
      const response = await client.graph.getContext(graphUuid, {
        query: request.body.query,
        ...(request.body.filters !== undefined ? { filters: request.body.filters } : {}),
      });
      const context = response.context?.trim();
      return context ? [context] : [];
    }
    case "edges": {
      const page = await client.graph.searchEdges(graphUuid, request);
      return page.data.map((edge) => edge.fact).filter(isNonEmpty);
    }
    case "nodes": {
      const page = await client.graph.searchNodes(graphUuid, request);
      return page.data.map(nameAndSummary).filter(isNonEmpty);
    }
    case "episodes": {
      const page = await client.graph.searchEpisodes(graphUuid, request);
      return page.data.map((episode) => episode.content).filter(isNonEmpty);
    }
    case "thread_summaries": {
      const page = await client.graph.searchThreadSummaries(graphUuid, request);
      return page.data.map((summary) => summary.summary).filter(isNonEmpty);
    }
    case "observations": {
      const page = await client.graph.searchObservations(graphUuid, request);
      return page.data.map(nameAndSummary).filter(isNonEmpty);
    }
    default: {
      const _exhaustive: never = scope;
      return _exhaustive;
    }
  }
}

/** Normalize `hiddenParams` (array or `Set`) into a `Set`. */
function toParamSet(
  value: Set<ZepSearchParamName> | ZepSearchParamName[] | undefined,
): Set<ZepSearchParamName> {
  return value instanceof Set ? value : new Set(value ?? []);
}

/**
 * Build a model-callable AI SDK tool that **searches** the bound Zep graph and
 * returns relevant facts.
 *
 * Drop it into `generateText`/`streamText`'s `tools` record so the model can
 * decide *when* and *what* to recall during a tool loop.
 *
 * **Pin-or-expose.** Every search parameter (`scope`, `reranker`, `limit`,
 * `mmrLambda`, `centerNodeUuid`) is exposed to the model in the tool's Zod
 * input schema by default. Use `pinnedParams` to fix a parameter to a constant
 * value and remove it from the schema (the model can no longer choose it); use
 * `hiddenParams` to remove a parameter from the schema *without* pinning it —
 * Zep's own server-side default applies, and the parameter is simply omitted
 * from the SDK call. `filters` and
 * `bfsOriginNodeUuids` are always constructor-only. The legacy `scope` /
 * `reranker` / `limit` constructor args pin (and thus hide) their parameter,
 * same as passing it via `pinnedParams` — back-compat for the pre-pin-or-expose
 * API.
 *
 * A Zep failure is logged (no PII) and returned as `found: false` with an empty
 * list; it never throws.
 */
export function createZepSearchTool(options: ZepSearchToolOptions) {
  const { client, binding } = options;
  const logger = resolveLogger(options.logger);
  const target = resolveGraphTarget(binding);

  const pinnedValues: Partial<Record<ZepSearchParamName, unknown>> = {
    ...options.pinnedParams,
  };
  // Legacy constructor args pin (and thus hide) their parameter.
  if (options.scope !== undefined) pinnedValues.scope ??= options.scope;
  if (options.reranker !== undefined) pinnedValues.reranker ??= options.reranker;
  if (options.limit !== undefined) pinnedValues.limit ??= options.limit;

  // Clamp a pinned limit to Zep's ceiling at construction so the call never
  // 400s.
  if (typeof pinnedValues.limit === "number") {
    pinnedValues.limit = clampSearchLimit(pinnedValues.limit, (m) => logger.warn(m));
  }

  // Auto scope rejects node_distance/episode_mentions and ignores reranker
  // entirely. If scope is pinned to "auto" and reranker is also pinned,
  // resolve the conflict once, here, so the call path is always valid.
  if (pinnedValues.scope === "auto" && "reranker" in pinnedValues) {
    if (AUTO_INCOMPATIBLE_RERANKERS.has(String(pinnedValues.reranker))) {
      logger.warn(
        `[zep-search] reranker '${String(pinnedValues.reranker)}' is invalid for ` +
          "scope 'auto'; omitting reranker (auto search uses RRF).",
      );
    }
    delete pinnedValues.reranker;
  }

  const pinned = new Set(Object.keys(pinnedValues) as ZepSearchParamName[]);
  const hidden = toParamSet(options.hiddenParams);

  const inputSchema = buildSearchInputSchema(pinned, hidden);

  return tool({
    description:
      options.description ??
      "Search long-term memory for facts about the user or domain learned in " +
        "previous turns or conversations. Use this to recall specific details " +
        "the user shared before.",
    inputSchema,
    execute: async (input: Record<string, unknown>): Promise<{ facts: string[]; found: boolean }> => {
      const trimmed = typeof input.query === "string" ? input.query.trim() : "";
      if (!trimmed) return { facts: [], found: false };
      if (!target) {
        logger.warn("[zep-search] No graphUuid bound; skipping search.");
        return { facts: [], found: false };
      }

      // Resolve each pin-or-exposable parameter: pinned > model-supplied >
      // Zep's own default (omitted from the call so the server applies it) —
      // except `scope`, which we default to "edges" client-side so
      // `extractResults` always knows which branch to read.
      const resolved: Partial<Record<ZepSearchParamName, unknown>> = {};
      for (const name of Object.keys(SEARCH_FIELD_BUILDERS) as ZepSearchParamName[]) {
        if (name in pinnedValues) {
          resolved[name] = pinnedValues[name];
        } else if (hidden.has(name)) {
          continue; // hidden, not pinned -> omit; Zep applies its own default
        } else if (input[name] !== undefined && input[name] !== null) {
          resolved[name] = input[name];
        }
      }

      const effectiveScope = (resolved.scope as ZepSearchScope | undefined) ?? DEFAULT_SEARCH_SCOPE;

      // Clamp a model-provided limit to Zep's ceiling — clamp, never reject:
      // the tool must not 400 on limit. (A pinned limit was already clamped at
      // construction, so this is a no-op for it.)
      if (typeof resolved.limit === "number") {
        resolved.limit = clampSearchLimit(resolved.limit, (m) => logger.warn(m));
      }

      // Auto search always uses RRF internally and ignores reranker entirely;
      // Zep rejects node_distance/episode_mentions outright. Drop any
      // (model-provided) reranker; warn only when Zep would have rejected it.
      if (effectiveScope === "auto" && "reranker" in resolved) {
        const droppedReranker = resolved.reranker;
        delete resolved.reranker;
        if (AUTO_INCOMPATIBLE_RERANKERS.has(String(droppedReranker))) {
          logger.warn(
            `[zep-search] reranker '${String(droppedReranker)}' is invalid for ` +
              "scope 'auto'; omitting reranker.",
          );
        }
      }

      try {
        // None/undefined-omission guard: only send a key when a value is
        // actually resolved — never an explicit null/undefined. `scope` selects
        // the method and `limit` is the page size, so neither belongs in the
        // search body.
        const bodyFields: Record<string, unknown> = { query: trimmed };
        for (const [name, value] of Object.entries(resolved)) {
          if (name === "scope" || name === "limit") continue;
          if (value !== undefined && value !== null) {
            bodyFields[name] = value;
          }
        }
        if (options.filters !== undefined) {
          bodyFields.filters = options.filters;
        }
        if (options.bfsOriginNodeUuids !== undefined) {
          bodyFields.bfsOriginNodeUuids = options.bfsOriginNodeUuids;
        }
        const request = {
          ...(typeof resolved.limit === "number" ? { limit: resolved.limit } : {}),
          body: bodyFields as unknown as Zep.SearchRequest,
        };

        const facts = await searchByScope(client, target, effectiveScope, request);
        return { facts, found: facts.length > 0 };
      } catch (error) {
        logger.warn(`[zep-search] Zep graph search failed: ${errorMessage(error)}`);
        return { facts: [], found: false };
      }
    },
  });
}

/**
 * Build a model-callable AI SDK tool that **persists** information into Zep.
 *
 * Conversational content (a `role` is provided and a thread is bound) is
 * written with `thread.addMessages`, which records history and ingests into the
 * user graph. Everything else is ingested with `graph.episode.add`
 * (`type: "text"`).
 *
 * Zep ingestion is **asynchronous**: a just-stored fact is not instantly
 * retrievable. The tool reports success once Zep accepts the data.
 *
 * Over-long content is truncated (lengths-only warning, never content). A Zep
 * failure is logged and surfaced as `stored: false`; it never throws.
 */
export function createZepRememberTool(options: ZepRememberToolOptions) {
  const { client, binding } = options;
  const logger = resolveLogger(options.logger);
  const target = resolveGraphTarget(binding);
  const threadUuid = binding.threadUuid;

  return tool({
    description:
      options.description ??
      "Persist a fact, preference, or message to long-term memory so it can be " +
        "recalled in future turns and future conversations.",
    inputSchema: rememberInputSchema,
    execute: async ({
      content,
      role,
    }): Promise<{ stored: boolean; message: string }> => {
      const trimmed = content?.trim();
      if (!trimmed) {
        return { stored: false, message: "Nothing to remember: content was empty." };
      }
      if (!target) {
        logger.warn("[zep-remember] No graphUuid bound; skipping persist.");
        return {
          stored: false,
          message: "Memory is not configured (no graph bound).",
        };
      }

      try {
        // Conversational content with a bound thread → thread.addMessages.
        // Capped at Zep's 4,096-char message limit.
        const roleType = toRoleType(role, logger);
        if (role && threadUuid) {
          await client.thread.addMessages(threadUuid, {
            messages: [
              {
                ...(roleType !== undefined ? { role: roleType } : {}),
                content: truncateForZep(trimmed, MESSAGE_MAX_CHARS, "zep-remember", logger),
                ...(options.defaultMessageName !== undefined
                  ? { name: options.defaultMessageName }
                  : {}),
              },
            ],
          });
          return { stored: true, message: "Saved to conversation memory." };
        }

        // Everything else → graph.episode.add as text. Capped at Zep's
        // 10,000-char episode limit (never silently dropped — truncate with a
        // warning).
        await client.graph.episode.add(target, {
          type: "text",
          data: truncateForZep(trimmed, GRAPH_MAX_CHARS, "zep-remember", logger),
        });
        return { stored: true, message: "Saved to long-term memory." };
      } catch (error) {
        logger.warn(`[zep-remember] Failed to persist to Zep: ${errorMessage(error)}`);
        return {
          stored: false,
          message: "Could not save to memory right now; continuing without it.",
        };
      }
    },
  });
}

/**
 * Build a model-callable AI SDK tool that returns the **Context Block** for the
 * bound user via `thread.getContext`.
 *
 * One call returns a prompt-ready string (user summary + relevant facts and
 * entities) assembled from the *whole* user graph, with the thread's most recent
 * messages used to focus relevance. It is the tool-callable counterpart to
 * {@link createZepMiddleware}'s system-message injection — useful when you want
 * the model to pull in context on demand inside a tool loop.
 *
 * A Zep failure is logged (no PII) and returned as `found: false` with an empty
 * string; it never throws.
 */
export function createZepContextTool(options: ZepContextToolOptions) {
  const { client, threadUuid } = options;
  const logger = resolveLogger(options.logger);

  return tool({
    description:
      options.description ??
      "Retrieve everything currently known about the user — a summary plus " +
        "relevant facts from previous conversations — to ground your response.",
    inputSchema: contextInputSchema,
    execute: async (): Promise<{ context: string; found: boolean }> => {
      if (!threadUuid) {
        logger.warn("[zep-context] No threadUuid bound; skipping context retrieval.");
        return { context: "", found: false };
      }
      try {
        const response = await client.thread.getContext(
          threadUuid,
          options.templateUuid ? { templateUuid: options.templateUuid } : {},
        );
        const context = response.context?.trim() ?? "";
        return { context, found: context.length > 0 };
      } catch (error) {
        logger.warn(`[zep-context] Failed to retrieve Zep context: ${errorMessage(error)}`);
        return { context: "", found: false };
      }
    },
  });
}

/** Options for {@link createZepTools}. */
export interface ZepToolsOptions extends Omit<BaseToolOptions, "client"> {
  /**
   * Graph + thread binding. For a personalized conversational agent supply the
   * user's `graphUuid` and the `threadUuid`; for a shared knowledge base supply
   * the standalone `graphUuid`. `zepContext` requires a `threadUuid`; without
   * one it returns a graceful empty result.
   */
  binding: ZepBinding & { threadUuid?: string };
  /** Pin the search scope (default `"edges"`). */
  searchScope?: ZepSearchScope;
  /** Pin the search result limit. */
  searchLimit?: number;
  /** Default speaker name recorded on conversational messages persisted by `zepRemember`. */
  defaultMessageName?: string;
}

/**
 * The Zep tool set, keyed for direct spread into a `generateText`/`streamText`
 * `tools` record:
 *
 * ```ts
 * const tools = createZepTools(client, { binding: { graphUuid, threadUuid } });
 * await generateText({ model, prompt, tools, stopWhen: stepCountIs(5) });
 * ```
 */
export type ZepTools = ToolSet & {
  /** Search the bound graph for relevant facts (`graph.search*`). */
  zepSearch: ReturnType<typeof createZepSearchTool>;
  /** Persist a message or fact (`thread.addMessages` / `graph.episode.add`). */
  zepRemember: ReturnType<typeof createZepRememberTool>;
  /** Retrieve the whole-user-graph Context Block (`thread.getContext`). */
  zepContext: ReturnType<typeof createZepContextTool>;
};

/**
 * Build the full set of Zep tools bound to a single client and binding.
 *
 * Spread the result into a `generateText`/`streamText` `tools` record. Each tool
 * handles Zep failures gracefully and never throws — a memory outage cannot
 * crash the host call. `zepContext` is included only when a `threadUuid` is
 * bound (it has no meaning without one).
 *
 * @param client - A shared, initialized Zep client.
 * @param options - The binding plus optional search/persist configuration.
 */
export function createZepTools(client: ZepClient, options: ZepToolsOptions): ZepTools {
  const { binding } = options;
  const logger = resolveLogger(options.logger);

  const tools = {
    zepSearch: createZepSearchTool({
      client,
      binding,
      ...(options.searchScope !== undefined ? { scope: options.searchScope } : {}),
      ...(options.searchLimit !== undefined ? { limit: options.searchLimit } : {}),
      logger,
    }),
    zepRemember: createZepRememberTool({
      client,
      binding,
      ...(options.defaultMessageName !== undefined
        ? { defaultMessageName: options.defaultMessageName }
        : {}),
      logger,
    }),
    zepContext: createZepContextTool({
      client,
      threadUuid: binding.threadUuid ?? "",
      logger,
    }),
  };
  return tools as ZepTools;
}
